ctapipe/tools/muon_reconstruction.py

```diff
@@ -5,121 +5,176 @@
+
+        mask = clean_mask
+        for i in range(3):
+            ring = self.ring_fitter(x, y, image, mask)
+            self.log.debug(
+                f'It {i}: r={ring.ring_radius:.2f}'
+                f', x={ring.ring_center_x:.2f}'
+                f', y={ring.ring_center_y:.2f}'
+            )
+
+            dist = np.sqrt((x - ring.ring_center_x)**2 + (y - ring.ring_center_y)**2)
+            mask = np.abs(dist - ring.ring_radius) / ring.ring_radius < 0.4
+
+        if np.count_nonzero(mask) <= self.min_pixels.tel[tel_id]:
+            self.log.info(
+                f'Skipping event {event_id}-{tel_id}:'
+                f' Less then {self.min_pixels.tel[tel_id]} pixels on ring'
+            )
+            return
+
+        if np.isnan([ring.ring_radius.value, ring.ring_center_x.value, ring.ring_center_y.value]).any():
+            self.log.info(f'Skipping event {event_id}-{tel_id}: Ring fit did not succeed')
+            return
+
+        # intensity_fitter does not support a mask yet, set ignored pixels to 0
+        image[~mask] = 0
+
+        result = self.intensity_fitter(
+            tel_id,
+            ring.ring_center_x,
+            ring.ring_center_y,
+            ring.ring_radius,
+            image,
+            pedestal=1.1,
+        )
 
-        numev = 0
-        self.num_muons_found = defaultdict(int)
-
-        for event in tqdm(self.source, desc='detecting muons'):
-
-            self.calib(event)
-            muon_evt = analyze_muon_event(event)
-
-            if numev == 0:
-                _exclude_some_columns(event.inst.subarray, self.writer)
-
-            numev += 1
-
-            if not muon_evt['MuonIntensityParams']:
-                # No telescopes  contained a good muon
-                continue
-            else:
-                if self.display:
-                    plot_muon_event(event, muon_evt)
-
-                for tel_id in muon_evt['TelIds']:
-                    idx = muon_evt['TelIds'].index(tel_id)
-                    intens_params = muon_evt['MuonIntensityParams'][idx]
+        self.log.info(
+            f'Muon fit: r={ring.ring_radius:.2f}'
+            f', width={result.ring_width:.4f}'
+            f', efficiency={result.optical_efficiency_muon:.2%}',
+        )
 
-                    if intens_params is not None:
-                        ring_params = muon_evt['MuonRingParams'][idx]
-                        cam_id = str(event.inst.subarray.tel[tel_id].camera)
-                        self.num_muons_found[cam_id] += 1
-                        self.log.debug("INTENSITY: %s", intens_params)
-                        self.log.debug("RING: %s", ring_params)
-                        self.writer.write(table_name=cam_id,
-                                          containers=[intens_params,
-                                                      ring_params])
+        if self.outfile:
+            self.writer.write(f'tel_{tel_id}', [event_index, ring, result])
```

Comment:
use `f'tel_{tel_id:03d}` so that the sort order is correct (tel_001, tel_002, ..tel_015, ... tel_152)